prov/mxm/src/mlxm.h

```diff
@@ -0,0 +1,376 @@
+#endif
+        union{
+                mxm_mq_h mq;
+                uint64_t is_set;
+                // TODO: static assert those are equal
+        };
+};
+
+
+
+struct mlxm_mq_storage{
+#if MLXM_MQ_STORAGE_TYPE == MLXM_MQ_STORAGE_LIST
+        struct dlist_entry mq_list;
+#elif MLXM_MQ_STORAGE_TYPE == MLXM_MQ_STORAGE_ARRAY
+        /* MXM mq id is uint16_t so the max number of different MQs
+           is 1 << 16. */
+        struct mlxm_mq_entry mq_entries[1<<16];
+#endif
+};
+
+
+struct mlxm_globals{
+        mxm_h mxm_context;
+        mxm_ep_h mxm_ep;
+        struct mlxm_mq_storage mq_storage;
+    struct mpool *req_pool;
+};
+extern struct mlxm_globals mlxm_globals;
+
+static inline
+int mlxm_find_mq(struct mlxm_mq_storage *storage,
+                 uint16_t id, mxm_mq_h *mq) {
+#if MLXM_MQ_STORAGE_TYPE == MLXM_MQ_STORAGE_LIST
+        struct dlist_entry *p, *head;
+        struct mlxm_mq_entry *mq_entry;
+
+        head = &storage->mq_list;
+
+        for (p = head->next; p != head; p = p->next) {
+                mq_entry = (struct mlxm_mq_entry*)p;
+                if (mq_entry->mq_key == id) {
+                        *mq = mq_entry->mq;
+                        return 0;
+                }
+        }
+#elif MLXM_MQ_STORAGE_TYPE == MLXM_MQ_STORAGE_ARRAY
+        if (storage->mq_entries[id].is_set != 0) {
+                *mq = storage->mq_entries[id].mq;
+                return 0;
+        }
+#endif
+        return 1;
+
+};
+
+static inline
+int mlxm_mq_add_to_storage(struct mlxm_mq_storage *storage,
+                           uint16_t id, mxm_mq_h *mq) {
+        int mxm_err;
+        struct mlxm_mq_entry *mq_entry;
+#if MLXM_MQ_STORAGE_TYPE == MLXM_MQ_STORAGE_LIST
+        mq_entry = (struct mlxm_mq_entry*)malloc(sizeof(*mq_entry));
+#elif MLXM_MQ_STORAGE_TYPE == MLXM_MQ_STORAGE_ARRAY
+        mq_entry = &storage->mq_entries[id];
+#endif
+
+        mxm_err = mxm_mq_create(mlxm_globals.mxm_context,
+				id,
+                                &mq_entry->mq);
+
+        if (mxm_err) {
+                FI_WARN(&mlxm_prov,FI_LOG_CORE,
+                        "mxm_mq_create failed: mq_id %d, errno %d:%s\n",
+                        id, mxm_err, mxm_error_string(mxm_err));
+                return mlxm_errno(mxm_err);
+	}
+
+        FI_INFO(&mlxm_prov,FI_LOG_CORE,
+                "MXM mq created, id 0x%x, %p\n",id , mq_entry->mq);
+
+#if MLXM_MQ_STORAGE_TYPE == MLXM_MQ_STORAGE_LIST
+                mq_entry->mq_key = id;
+                dlist_insert_after((struct dlist_entry *)mq_entry,
+                                   (struct dlist_entry *)storage);
+#endif
+
+        *mq = mq_entry->mq;
+        return 0;
+};
+
+static inline
+void mlxm_mq_storage_init() {
+#if MLXM_MQ_STORAGE_TYPE == MLXM_MQ_STORAGE_LIST
+    dlist_init(&mlxm_globals.mq_storage.mq_list);
+#elif MLXM_MQ_STORAGE_TYPE == MLXM_MQ_STORAGE_ARRAY
+    memset(mlxm_globals.mq_storage.mq_entries,0,
+                sizeof(mlxm_globals.mq_storage.mq_entries));
+#endif
+}
```

Comment:
There are a lot of #if MLXM_MQ_STORAGE_TYPE checks.  These make the code hard to read and are a compile time check.  What purpose do these serve?  Do these need to be compile options?